### PR TITLE
Feat/observability support

### PR DESCRIPTION
**Pull request by Pouyanpi**
## Description
Implements a tracing module to assign a trace id to every incoming request and enable detailed tracing of every rail/action/LLM call that is triggered behind the scenes. Can export the traces to different providers like OpenTelemetry, etc.

## Example Usage

Consider following `config.yml`

```yaml
models:
  - type: main
    engine: openai
    model: gpt-3.5-turbo-instruct

tracing:
  enabled: true
  adapters:
    - name: OpenTelemetry
      service_name: "nemo_guardrails_service"
      exporter: "console"
      resource_attributes:
        env: "production"
    - name: FileSystem
      filepath: "./traces/traces.jsonl"
    
```

Now you can also use `zipkin` as expoerter.

To do so:

```zsh
pip install opentelemetry-exporter-zipkin

```
```zsh
docker run -d -p 9411:9411 openzipkin/zipkin
```
and set the exporter to zipkin in config.yml :
```yaml
tracing:
  enabled: true
  adapters:
    - name: OpenTelemetry
      service_name: "nemo_guardrails_service"
      exporter: "zipkin"
      resource_attributes:
        env: "production"
   
```

## Note:
- `adapter_factory` is temporary, it will be replaced by a Registry pattern.

- Testing it in Colang 2.x depends on #797 

## Test Coverage

| File                                      | Statements | Missed | Coverage | Missing Lines |
|-------------------------------------------|------------|--------|----------|----------------|
| tests/test_tracing.py                     | 49         | 1      | 98%      | 128            |
| tests/test_tracing_adapters_filesystem.py | 31         | 0      | 100%     | -              |
| tests/test_tracing_adapters_opentelemetry.py | 69         | 5      | 93%      | 146-152, 186   |

## Checklist

- [x] I've read the [CONTRIBUTING](https://github.com/NVIDIA/NeMo-Guardrails/blob/develop/CONTRIBUTING.md) guidelines.
- [x] I've updated the documentation if applicable.
- [x] I've added tests if applicable.
- [x] opentelemetry dependencies in pyproject.toml - @drazvan

### CODE REVIEW

File: nemoguardrails/rails/llm/config.py
Comment (drazvan) [inline nemoguardrails/rails/llm/config.py]: Maybe rename to `LogAdapterConfig`

File: nemoguardrails/rails/llm/llmrails.py
Comment (drazvan) [inline nemoguardrails/rails/llm/llmrails.py]: This seems blocking.

File: nemoguardrails/tracing/adapters/filesystem.py
Comment (drazvan) [inline nemoguardrails/tracing/adapters/filesystem.py]: This should be .jsonl.

File: nemoguardrails/rails/llm/llmrails.py
Comment (drazvan) [inline nemoguardrails/rails/llm/llmrails.py:909]: To check if this initializes OTEL every time or not.

**Review by drazvan — CHANGES_REQUESTED** [review]
Looks good :+1:. After fixing the issues, good to merge.

**Comment by Pouyanpi** [comment]
@drazvan the changes are done. Currently tracing dependencies are added to `pyproject.toml` and using defered import like similar situations in the codebase would clutter the `adapters.opentelmetry` module. Dependency req is included in the documentation.

**Review by drazvan — APPROVED** [review]
Approving to unblock. Will review in more detail next week.

**Pull request merged**
